src/hotspot/share/services/memReporter.hpp

```diff
@@ -71,11 +71,29 @@ class MemReporterBase : public StackObj {
   }
 
   // Convert diff amount in bytes to current reporting scale
-  inline long diff_in_current_scale(size_t s1, size_t s2) const {
-    long amount = (long)(s1 - s2);
-    long scale = (long)_scale;
-    amount = (amount > 0) ? (amount + scale / 2) : (amount - scale / 2);
-    return amount / scale;
+  int64_t diff_in_current_scale(size_t s1, size_t s2) const {
+    // _scale should not be 0, otherwise division by zero at return.
+    assert(_scale != 0, "wrong scale");
+
+    bool is_negative = false;
+    if (s1 < s2) {
+      is_negative = true;
+      swap(s1, s2);
+    }
+
+    size_t amount = s1 - s2;
+    assert(amount <= SIZE_MAX - _scale / 2, "size_t overflow");
+    size_t scaled = (amount / _scale);
+    if ((amount % _scale) > (_scale - 1)/2) {
+      scaled += 1;
+    }
+
+    if (scaled == 0) {
```

Comment:
I don't think this special case is needed.  It falls out from the generic code, and is likely pretty unusual.

Reply:
It's not needed because there is the assert above checking it's not 0.